tests/cnf/ran/powermanagement/internal/collect/collect.go

```diff
@@ -0,0 +1,266 @@
+package collect
+
+import (
+	"fmt"
+	"math"
+	"slices"
+	"strconv"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/openshift-kni/eco-goinfra/pkg/nto" //nolint:misspell
+	"github.com/openshift-kni/eco-goinfra/pkg/pod"
+	. "github.com/openshift-kni/eco-gotests/tests/cnf/ran/internal/raninittools"
+	"github.com/openshift-kni/eco-gotests/tests/cnf/ran/internal/stats"
+	"github.com/openshift-kni/eco-gotests/tests/cnf/ran/powermanagement/internal/tsparams"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/utils/cpuset"
+)
+
+// GetPowerState determines the power state from the workloadHints object of the PerformanceProfile.
+func GetPowerState(perfProfile *nto.Builder) (string, error) {
+	workloadHints := perfProfile.Definition.Spec.WorkloadHints
+
+	if workloadHints == nil {
+		// No workloadHints object -> default is Performance mode
+		return tsparams.PerformanceMode, nil
+	}
+
+	realTime := *workloadHints.RealTime
+	highPowerConsumption := *workloadHints.HighPowerConsumption
+	perPodPowerManagement := *workloadHints.PerPodPowerManagement
+
+	switch {
+	case realTime && !highPowerConsumption && !perPodPowerManagement:
+		return tsparams.PerformanceMode, nil
+	case realTime && highPowerConsumption && !perPodPowerManagement:
+		return tsparams.HighPerformanceMode, nil
+	case realTime && !highPowerConsumption && perPodPowerManagement:
+		return tsparams.PowerSavingMode, nil
+	default:
+		return "", fmt.Errorf("unknown workloadHints power state configuration")
+	}
+}
+
+// CollectPowerMetricsWithNoWorkload collects metrics with no workload.
+func CollectPowerMetricsWithNoWorkload(duration, interval time.Duration, tag string) (map[string]string, error) {
+	glog.V(tsparams.LogLevel).Infof("Wait for %s for noworkload scenario", duration)
+
+	return collectPowerUsageMetrics(duration, interval, "noworkload", tag)
+}
+
+// CollectPowerMetricsWithSteadyWorkload collects power metrics with steady workload scenario.
+func CollectPowerMetricsWithSteadyWorkload(
+	duration, interval time.Duration, tag string, perfProfile *nto.Builder, nodeName string) (map[string]string, error) {
+	// stressNg cpu count is roughly 75% of total isolated cores.
+	// 1 cpu will be used by other consumer pods, such as process-exporter, cnf-ran-gotests-priv.
+	isolatedCPUSet, err := cpuset.Parse(string(*perfProfile.Object.Spec.CPU.Isolated))
+	if err != nil {
+		return nil, err
+	}
+
+	stressNgCPUCount := (isolatedCPUSet.Size() - 1) * 300 / 400
+	stressngMaxPodCount := 50
+
+	stressNgPods, err := deployStressNgPods(stressNgCPUCount, stressngMaxPodCount, nodeName)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(stressNgPods) < 1 {
+		return nil, fmt.Errorf("not enough stress-ng pods to run test")
+	}
+
+	glog.V(tsparams.LogLevel).Infof("Wait for %s for steadyworkload scenario", duration.String())
+	result, collectErr := collectPowerUsageMetrics(duration, interval, "steadyworkload", tag)
```

Comment:
Should `collectErr` be checked immediately after this?

Reply:
I think this came up when the test got merged the first time lol but my idea is that we want to tear down the stress pods regardless of whether the metrics collection failed. Might be better to delete the pods in a `defer` but the preference seems to be against `defer` in tests

Reply:
I think `defer` is fine in tests just outside of a loop.

Sounds good though!